### PR TITLE
i18n #109

### PR DESCRIPTION
**Pull request by tfranzel**
i18n support for build-in strings, CLI command, schema endpoint and UI endpoints.

caveat: of course the user has to provide his own translations. example translation only applies to the tests.

@jayvdb please review changes and functionality.

### CODE REVIEW

File: drf_spectacular/views.py [hunk 1]
```diff
@@ -42,9 +51,15 @@ def get(self, request, *args, **kwargs):
...
+            with translation.override(request.GET.get('lang')):
+                return self._get_schema_response(request)
+        else:
+            return self._get_schema_response(request)
+
+    def _get_schema_response(self, request):
         generator = self.generator_class(urlconf=self.urlconf, api_version=self.api_version)
-        schema = generator.get_schema(request=request, public=self.serve_public)
-        return Response(schema)
+        return Response(generator.get_schema(request=request, public=self.serve_public))
```
Comment (jayvdb) [inline drf_spectacular/views.py:62]: unrelated/unnecessary change here, if that matters to you, and didnt have a good reason for it.

File: drf_spectacular/views.py [hunk 2]
```diff
@@ -64,8 +79,11 @@ class SpectacularSwaggerView(APIView):
 
     @extend_schema(exclude=True)
     def get(self, request, *args, **kwargs):
+        schema_url = self.url or reverse(self.url_name, request=request)
+        if request.GET.get('lang'):
+            schema_url += f'?lang={request.GET.get("lang")}'
```
Comment (jayvdb) [inline drf_spectacular/views.py:84]: `?` may already be in schema_url; if so, need to use `&` instead

File: drf_spectacular/views.py [hunk 3]
```diff
@@ -79,7 +97,10 @@ class SpectacularRedocView(APIView):
...
```
Comment (jayvdb) [inline drf_spectacular/views.py:102]: as above

File: drf_spectacular/openapi.py [hunk 4]
```diff
@@ -818,7 +819,7 @@ def _get_response_bodies(self):
 
         if is_serializer(response_serializers) or is_basic_type(response_serializers):
             if self.method == 'DELETE':
-                return {'204': {'description': 'No response body'}}
+                return {'204': {'description': str(_('No response body'))}}
```
Comment (jayvdb) [inline drf_spectacular/openapi.py:822]: no translation for this yet.  As it is likely to be the first & most common string emitted by DRF-spectacular, it would be nice to have this one done.  "Unspecified .." also
But if not in this PR, I'll get it done.

File: drf_spectacular/views.py [hunk 5]
```diff
@@ -11,10 +13,17 @@
...
-from drf_spectacular.utils import extend_schema
+from drf_spectacular.utils import OpenApiParameter, extend_schema
 
 if spectacular_settings.SERVE_INCLUDE_SCHEMA:
     SCHEMA_KWARGS: Dict[str, Any] = {'responses': {200: OpenApiTypes.OBJECT}}
+
+    if settings.USE_I18N:
+        SCHEMA_KWARGS['parameters'] = [
+            OpenApiParameter(
+                'lang', str, OpenApiParameter.QUERY, enum=list(dict(settings.LANGUAGES).keys())
```
Comment (jayvdb) [inline drf_spectacular/views.py:24]: sort the lang keys?

currently I get the same order as my setting, which is not alpha ordered

**Comment by jayvdb** [comment]
When I try settings

```py
SPECTACULAR_SETTINGS = {
    'TITLE': _('...'),
    'DESCRIPTION': _('...'),
    ...
}
```

I get

```py
  File "/usr/lib/python3.8/site-packages/django/core/management/base.py", line 369, in execute
    output = self.handle(*args, **options)
  File "/usr/lib/python3.8/site-packages/drf_spectacular/management/commands/spectacular.py", line 60, in handle
    output = renderer.render(schema, renderer_context={})
  File "/usr/lib/python3.8/site-packages/drf_spectacular/renderers.py", line 33, in render
    return yaml.dump(
  File "/usr/lib64/python3.8/site-packages/yaml/__init__.py", line 290, in dump
    return dump_all([data], stream, Dumper=Dumper, **kwds)
  File "/usr/lib64/python3.8/site-packages/yaml/__init__.py", line 278, in dump_all
    dumper.represent(data)
  File "/usr/lib64/python3.8/site-packages/yaml/representer.py", line 27, in represent
    node = self.represent_data(data)
  File "/usr/lib64/python3.8/site-packages/yaml/representer.py", line 48, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/usr/lib64/python3.8/site-packages/yaml/representer.py", line 207, in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
  File "/usr/lib64/python3.8/site-packages/yaml/representer.py", line 118, in represent_mapping
    node_value = self.represent_data(item_value)
  File "/usr/lib64/python3.8/site-packages/yaml/representer.py", line 48, in represent_data
    node = self.yaml_representers[data_types[0]](self, data)
  File "/usr/lib64/python3.8/site-packages/yaml/representer.py", line 207, in represent_dict
    return self.represent_mapping('tag:yaml.org,2002:map', data)
  File "/usr/lib64/python3.8/site-packages/yaml/representer.py", line 118, in represent_mapping
    node_value = self.represent_data(item_value)
  File "/usr/lib64/python3.8/site-packages/yaml/representer.py", line 58, in represent_data
    node = self.yaml_representers[None](self, data)
  File "/usr/lib64/python3.8/site-packages/yaml/representer.py", line 231, in represent_undefined
    raise RepresenterError("cannot represent an object", data)
yaml.representer.RepresenterError: ('cannot represent an object', '<German ....(must longer of course)....>')
```

File: drf_spectacular/views.py (hunk 2 above)
Comment (tfranzel) [reply drf_spectacular/views.py]: done

File: drf_spectacular/views.py (hunk 3 above)
Comment (tfranzel) [reply drf_spectacular/views.py]: done

File: drf_spectacular/views.py (hunk 5 above)
Comment (tfranzel) [reply drf_spectacular/views.py:25]: opted against. imho the user should decide which order is important to him.

File: drf_spectacular/views.py (hunk 1 above)
Comment (tfranzel) [reply drf_spectacular/views.py:63]: no particular reason. just looked cleaner to me in that spot

File: drf_spectacular/openapi.py (hunk 4 above)
Comment (tfranzel) [reply drf_spectacular/openapi.py]: i prefer not to ship language files as then we would do a whole lot of languages. if someone cares for those  3-4 strings they can translate them when they do their app anyway.

**Comment by tfranzel** [comment]
> tests/settings.py appearing in the github diff viewer 

that was on purpose. test setup is sometimes picky about those files.

> 'TITLE': _('English title'),

that should work now. the whole schema is parsed for lazy objects now. so all representer error should be gone now.

> description: "Je gr\xF6\xDFer die Zahl, desto weiter oben taucht das Land
\ in der Liste auf."

the output should be unicode/utf-8 everywhere now. there was an option missing. added YAML test file where it looks good now. json tested locally and also looking good.

> dogfood, yum-yum

and now i ate it. encapulated the docstring

> description: A unique integer value identifying this Zeilenattribut.

actually it is working as expected. that string comes from DRF and they are missing the translation. upstream bug i suppose. looks like nobody has updated that file in years. https://github.com/encode/django-rest-framework/blob/master/rest_framework/locale/de/LC_MESSAGES/django.po


@jayvdb thanks for the thorough review. much appreciated. i think we covered everything. can you give it another quick pass and see if stuff works as expected?
